api/admin/controller.py

```diff
@@ -1032,3 +1042,81 @@ def collections(self):
             return Response(unicode(_("Success")), 201)
         else:
             return Response(unicode(_("Success")), 200)
+
+    def admin_auth_services(self):
+        if flask.request.method == 'GET':
+            auth_services = []
+            auth_service = get_one(self._db, AdminAuthenticationService)
+            if auth_service and auth_service.provider == AdminAuthenticationService.GOOGLE_OAUTH:
+                auth_services = [
+                    dict(
+                        name=auth_service.name,
+                        provider=auth_service.provider,
+                        url=auth_service.external_integration.url,
+                        username=auth_service.external_integration.username,
+                        password=auth_service.external_integration.password,
+                        domains=json.loads(auth_service.external_integration.setting("domains").value),
+                    )
+                ]
+
+            return dict(
+                admin_auth_services=auth_services,
+                providers=AdminAuthenticationService.PROVIDERS,
+            )
+
+        name = flask.request.form.get("name")
+        if not name:
+            return MISSING_ADMIN_AUTH_SERVICE_NAME
+
+        provider = flask.request.form.get("provider")
+
+        if provider and provider not in AdminAuthenticationService.PROVIDERS:
+            return UNKNOWN_ADMIN_AUTH_SERVICE_PROVIDER
+
+        is_new = False
+        auth_service = get_one(self._db, AdminAuthenticationService)
+        if auth_service:
+            # Currently there can only be one admin auth service, and one already exists.
+            if name != auth_service.name:
+                return ADMIN_AUTH_SERVICE_NOT_FOUND
+
+            if provider != auth_service.provider:
+                return CANNOT_CHANGE_ADMIN_AUTH_SERVICE_PROVIDER
+
+        else:
+            if provider:
+                auth_service, is_new = get_one_or_create(
+                    self._db, AdminAuthenticationService, name=name, provider=provider
+                )
+            else:
+                return NO_PROVIDER_FOR_NEW_ADMIN_AUTH_SERVICE
+
+        # Only Google OAuth is supported for now.
+        url = flask.request.form.get("url")
+        username = flask.request.form.get("username")
+        password = flask.request.form.get("password")
+        domains = flask.request.form.get("domains")
+        
+        if not url or not username or not password or not domains:
+            # If an admin auth service was created, make sure it
+            # isn't saved in a incomplete state.
+            self._db.rollback()
+            return INCOMPLETE_ADMIN_AUTH_SERVICE_CONFIGURATION
```

Comment:
I would expect this to be handled by `shutdown_session`. Is that not happening because this isn't an 'exception'?